pkg/cmd/task/task.go

```diff
@@ -32,9 +41,104 @@ func NewCommand(ctx api.Context) *cobra.Command {
 	cmd.Flags().Bool("completed", false, "Print completed tasks")
 
 	cmd.AddCommand(
+		newCmdTaskAttach(ctx),
 		newCmdTaskList(ctx),
 		newCmdTaskLog(ctx),
 		newCmdTaskLs(ctx),
 	)
 	return cmd
 }
+
+func findTask(ctx api.Context, id string) (*mesosgo.Task, error) {
+	tasks, err := findTasks(ctx, id)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(tasks) > 1 {
+		var names []string
+		for _, task := range tasks {
+			names = append(names, task.TaskID.Value)
+		}
+		return nil, fmt.Errorf("found more than one task with the same name: %v", names)
+	}
+
+	return &tasks[0], nil
+}
+
+func findTasks(ctx api.Context, id string) ([]mesosgo.Task, error) {
+	mesosClient, err := mesos.NewClientWithContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	allTasks, err := mesosClient.Tasks()
+	if err != nil {
+		return nil, err
+	}
+
+	g, err := glob.Compile(id)
+	if err != nil {
+		return nil, err
+	}
+
+	var tasks []mesosgo.Task
+	for _, t := range allTasks {
+		if strings.Contains(t.TaskID.Value, id) || g.Match(t.TaskID.Value) {
+			tasks = append(tasks, t)
+		}
+	}
+
+	if len(tasks) == 0 {
+		return nil, fmt.Errorf("no task ID found containing '%s'", id)
+	}
+	return tasks, nil
+}
+
+func mesosHTTPClient(ctx api.Context, agentID string) (*httpcli.Client, error) {
+	cluster, err := ctx.Cluster()
+	if err != nil {
+		return nil, err
+	}
+
+	mesosURL, _ := cluster.Config().Get("core.mesos_master_url").(string)
+	if mesosURL != "" {
+		mesosClient, err := mesos.NewClientWithContext(ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		agents, err := mesosClient.Agents()
+		if err != nil {
+			return nil, err
+		}
+
+		for _, a := range agents {
+			if a.AgentInfo.ID.Value == agentID {
+				// PID format: Host@IP:Port
+				s1 := strings.Split(a.GetPID(), "@")
+				s2 := strings.Split(s1[1], ":")
+				url := fmt.Sprintf("http://%s:%s", s2[0], s2[1])
+
+				return httpcli.New(
+					httpcli.Endpoint(url),
+
+					// Use JSON over protobuf to ease logging.
+					httpcli.Codec(codecs.ByMediaType[codecs.MediaTypeJSON]),
```

Comment:
This is mainly for debugging purpose, when users run the CLI in verbose mode they can then see HTTP request/response JSON bodies being printed, this is not possible with binary payloads.

Also, the Python codebase currently uses JSON as well. Do you have concrete examples of a call which wouldn't work?

Reply:
We had the issue once where Mesos added a key in one of its response and the CLI started to break, it was solved here: https://github.com/dcos/dcos-core-cli/commit/8017331eb64fba73bbca498569a768ae730288c9#diff-65145422e82c3e8089c40d2b75584a43
The Python codebase is not affected AFAIK because the issue happens when we are decoding the JSON response into a mesos-go struct.